src/Data/Variant.purs

```diff
@@ -73,6 +79,70 @@ on p f g r =
   coerceR ∷ Variant r2 → Variant r1
   coerceR = unsafeCoerce
 
+-- | Type class that matches a row for a `record` that will eliminate a row for
+-- | a `variant`, producing a `result` of the specified type. Just a wrapper for
+-- | `VariantMatch` to convert `RowToList` and vice versa.
+class VariantRecordElim
+    (variant ∷ # Type)
+    (record ∷ # Type)
+    result
+  | variant result → record
+  , record → variant result
```

Comment:
I think these fundeps should be `| variant -> record result, record -> variant result` as the instances (delegated to rowlists) only match on the variant and record - and only one of these is sufficient to infer the structure for the other.  We never match on `result`, so it can always determined.